inst/resources/lua/custom-environment.lua

```diff
@@ -0,0 +1,197 @@
+            print(label.." "..obj)
+        elseif type(obj) == "table" then
+            for k,v in iter(obj) do
+                print(label.."id:"..k.. " val:"..v)
+            end
+        end
+    end
+    return nil
+end
+
+-- create a unique id for a div with none provided
+local counter = 0
+function unlabeled_div()
+    counter = counter + 1 
+    return "unlabeled-div-"..(counter)
+end
+
+-- Get metadata specific to bookdown for this filter
+Meta = function(m) 
+    bookdownmeta = m.bookdown
+    if (bookdownmeta ~= nil) then
+        -- For internationalization feature of bookdown
+        if (bookdownmeta.language ~= nil) then
+            if (bookdownmeta.language.label ~= nil) then
+                for k,v in pairs(bookdownmeta.language.label) do
+                    if (type(v) == 'table' and v.t == 'MetaInlines' and proof_label[k] ~= nil) then
+                        -- remove any undesired space (3 or less)
+                        proof_label[k] = pandoc.utils.stringify(v):gsub("%.?%s?%s?%s?$", "")
+                        print_debug("Translation-> "..k..":", proof_label[k])
+                    end
+                end
+            end
+        end
+    end
+end
+
+-- Modify Pandoc AST for supported custom environment
+Div = function (div)
+    local classes = div.classes
+    -- we do nothing if no classes
+    if (#classes == 0) then
+        print_debug("No classes in the Div.")
+        return div
+    end
+    print_debug("Div classes -> " , classes)
+
+    -- checking if the class is one of the supported custom environment
+    local env_type = {type = nil, env = nil}
+    for i,v in ipairs(classes) do
+        if (theorem_abbr[v] ~= nil) then 
+            env_type.type = "theorem"
+            env_type.env = v
+            break
+        elseif (proof_label[v] ~= nil) then 
+            env_type.type = "proof"
+            env_type.env = v
+            break
+        end
+    end
+    -- classes is not a supported one, we return as is
+    if not env_type.env then
+        print_debug("Not a bookdown supported custom class")
+        return div
+    end
+    print_debug("Found types -> ", env_type)
+
+    -- get the id if it exists - it will we use to build label for reference
+    local id = div.identifier
+    -- if no id, one is generated so that bookdown labelling mechanism works
+    if #id == 0 then id = unlabeled_div() end
+    print_debug("id -> ", id)
+    -- remove unwanted identifier on the div, as it will be on the span
+    div.identifier = ""
+
+    -- get the attributes
+    local options = div.attributes
+    if (options["data-latex"] ~= nil) then 
+        -- so that latex-divs.lua in rmarkdown does not activate
+        print("[WARNING] data-latex attribute can't be used with one of bookdown custom environment. It has been removed.")
+        options["data-latex"] = nil
+    end
+
+    -- return [name] for latex, and (name) for html
+    function get_name(format, options)
```

Comment:
What about defining this function at the top level?
This function only uses local variables and given parameters.

```
local get_name = function(format, options)
...
```

Even if we need to define functions within a function, it is a good practice to define function locally like the above. Otherwise, `get_name` is defined globally and overwritten every time the `Div` function is called.

Reply:
I moved it out and define it as a local function in c45c9b867011f54012725119c74b0b55c74488e5